sway/desktop/transaction.c

```diff
@@ -606,21 +606,15 @@ static void arrange_output(struct sway_output *output, int width, int height) {
 	}
 }
 
-static void arrange_popup(struct wlr_scene_tree *popup) {
+void arrange_popups(struct wlr_scene_tree *popups) {
 	struct wlr_scene_node *node;
-	wl_list_for_each(node, &popup->children, link) {
-		struct sway_xdg_popup *popup = scene_descriptor_try_get(node,
+	wl_list_for_each(node, &popups->children, link) {
+		struct sway_popup_desc *popup = scene_descriptor_try_get(node,
 			SWAY_SCENE_DESC_POPUP);
 
-		// the popup layer may have popups from layer_shell surfaces, in this
-		// case those don't have a scene descriptor, so lets skip those here.
-		if (popup) {
-			struct wlr_scene_tree *tree = popup->view->content_tree;
-
-			int lx, ly;
-			wlr_scene_node_coords(&tree->node, &lx, &ly);
-			wlr_scene_node_set_position(&popup->scene_tree->node, lx, ly);
-		}
+		int lx, ly;
```

Comment:
I think this should contain a part for offset, maybe in `sway_popup_desc`, add a field of 

* bool has_offset
* int x_offset
* int y_offset

emm, what do you thought about?

Reply:
For input_method popups, I think it would just be easier to nest an extra scene tree. The position of the popup will be managed by this code here, and then you can apply your own offset to the nested tree. This can make things easier because if you want to change the offset later, you don't need to worry about clobbering the original position.

Reply:
I see, I should create a subsurface_tree of the view, set the position, and use this node as the relative